jcommune-service/src/main/java/org/jtalks/jcommune/service/transactional/TransactionalTopicDraftService.java

```diff
@@ -0,0 +1,81 @@
+ * version 2.1 of the License, or (at your option) any later version.
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package org.jtalks.jcommune.service.transactional;
+
+import org.jtalks.jcommune.model.dao.TopicDraftDao;
+import org.jtalks.jcommune.model.entity.*;
+import org.jtalks.jcommune.plugin.api.service.PluginTopicDraftService;
+import org.jtalks.jcommune.service.TopicDraftService;
+import org.jtalks.jcommune.service.UserService;
+import org.springframework.security.access.prepost.PreAuthorize;
+
+/**
+ * The implementation of TopicDraftService interface.
+ *
+ * @author Dmitry S. Dolzhenko
+ */
+public class TransactionalTopicDraftService implements TopicDraftService, PluginTopicDraftService {
+
+    private final UserService userService;
+    private final TopicDraftDao topicDraftDao;
+
+    public TransactionalTopicDraftService(UserService userService, TopicDraftDao topicDraftDao) {
+        this.userService = userService;
+        this.topicDraftDao = topicDraftDao;
+    }
+
+    @Override
+    public TopicDraft getDraft() {
+        JCUser user = userService.getCurrentUser();
+        return topicDraftDao.getForUser(user);
+    }
+
+    @Override
+    @PreAuthorize("( (not #topicDto.codeReview) and (not #topicDto.plugable) " +
+            "and hasPermission(#topicDto.branch.id, 'BRANCH', 'BranchPermission.CREATE_POSTS')) " +
+            "or (#topicDto.codeReview " +
+            "and hasPermission(#topicDto.branch.id, 'BRANCH', 'BranchPermission.CREATE_CODE_REVIEW')) " +
+            " or #topicDto.plugable") // User should have the same access rights as when creating a new topic
+    public TopicDraft saveOrUpdateDraft(Topic topicDto, String content) {
+        JCUser user = userService.getCurrentUser();
+
+        TopicDraft topicDraft = topicDraftDao.getForUser(user);
+        if (topicDraft == null) {
+            topicDraft = new TopicDraft(user, topicDto.getTitle(), content);
+        } else {
+            topicDraft.setContent(content);
+            topicDraft.setTitle(topicDto.getTitle());
+        }
+
+        Poll poll = topicDto.getPoll();
+        if (poll != null) {
+            topicDraft.setPollTitle(poll.getTitle());
+            topicDraft.setPollItemsValue(poll.getPollItemsValue());
+        }
+
+        topicDraftDao.saveOrUpdate(topicDraft);
+
+        return topicDraft;
+    }
+
+    @Override
+    public void deleteDraft() {
+        JCUser user = userService.getCurrentUser();
+        TopicDraft topicDraft = topicDraftDao.getForUser(user);
+        if (topicDraft != null) {
```

Comment:
If we add method `deleteDraft(User user)` to the DAO itself, it would allow us to delete it with one HQL `delete` instead of loading it first. Will be 2x faster :)